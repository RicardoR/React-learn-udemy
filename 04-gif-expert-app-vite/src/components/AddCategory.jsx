import { useState } from "react";

export const AddCategory = ({onNewCategory}) => {

    const [inputValue, setInputValue] = useState();

    const onInputChange = ({target}) => {
        setInputValue(target.value);
    };

    const onFormSubmit = (event) => {
        event.preventDefault();

        if (inputValue.trim().length <= 3) {
            return;
        }

        onNewCategory(inputValue.trim());
        setInputValue('');
    };

    return (
        <form onSubmit={onFormSubmit}>
            <input
                type="text"
                placeholder="Search gifs"
                value={inputValue}
                onChange={onInputChange}
            />
        </form>
    );
};

