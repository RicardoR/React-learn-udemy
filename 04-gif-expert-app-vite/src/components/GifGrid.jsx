import PropTypes from 'prop-types';
import { getGifs } from '../helpers/getGifs';


export const GifGrid = ({category}) => {

    getGifs(category);

    return (
        <>
            <h3>{category}</h3>
        </>
    );
};


GifGrid.propTypes = {
    category: PropTypes.string.isRequired
};
