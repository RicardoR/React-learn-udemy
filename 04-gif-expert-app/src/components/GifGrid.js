import React, { useEffect, useState } from 'react';
import { apiKey, giphySearchEndPoint } from '../secret/giphy-data';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

    const [imageList, setImages] = useState([]);

    useEffect(() => {
      getGifs();
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);
    
    const getGifs = async () => {
      const url = `${giphySearchEndPoint}?api_key=${apiKey}&q=${category}&limit=10`;
      const response = await fetch(url);
      const { data } = await response.json();
      
      const gifs = data.map(({ id, title, images }) => {
        return {
          id: id,
          title: title,
          url: images?.downsized_medium.url,
          };
        });
        setImages(gifs);
    };


  return (
    <>
      <h3>{category}</h3>
        {
        imageList.map((img) =>
          <GifGridItem
            {...img}
            key={img.id}
          />
        )
        }
    </>
  );
}
