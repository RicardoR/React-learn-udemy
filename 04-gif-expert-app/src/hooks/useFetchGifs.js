import { useState, useEffect } from 'react';
import { getGifs } from '../helpers/getGifs';

export const useFetchGifs = (category) => {
	const [state, setState] = useState({
		data: [],
		loading: true,
	});

	// useEffect only triggers when categroy changes
	useEffect(() => {
		getGifs(category).then((imgs) => {
			setTimeout(() => {
				setState({
					data: imgs,
					loading: false,
				});
			}, 1000);
		});
  }, [category]);

  return state;
};