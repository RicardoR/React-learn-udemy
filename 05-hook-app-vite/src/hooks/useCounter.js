import { useState } from 'react';

export const useCounter = (inititalValue = 10) => {
  const [counter, setCounter] = useState(inititalValue);

  const reset = () => {
    setCounter(inititalValue);
  };

  const increment = (value = 1) => {
    setCounter(counter + value);
  };

  const decrement = (value = 1) => {
    if (counter - value <= 0) {
      return;
    }
    setCounter(counter - 1);
  };
  return {
    counter,
    increment,
    decrement,
    reset,
  };
};
