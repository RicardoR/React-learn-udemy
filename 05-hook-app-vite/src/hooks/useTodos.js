import {useEffect, useReducer, useState} from 'react';
import {addTodoAction, deleteTodoAction, todoReducer, toggleTodoAction} from "../08-useReducer/todoReducer";

const LOCALSTORAGE_TODOS_KEY = 'todos';
const initialState = []

const init = () => {
    return JSON.parse(localStorage.getItem(LOCALSTORAGE_TODOS_KEY)) ?? [];
}

export const useTodos = () => {

    const [todos, dispatch] = useReducer(todoReducer, initialState, init);

    const [pendingTodosCount, setPendingTodosCount] = useState(0);
    const [todosCount, setTodosCount] = useState(0);

    useEffect(() => {
        const todosToStore = JSON.stringify(todos);
        localStorage.setItem(LOCALSTORAGE_TODOS_KEY, todosToStore);
        const pending = todos.filter(todo => todo.done === false)?.length;
        setPendingTodosCount(pending);
        setTodosCount(todos.length);
    }, [todos]);

    const handleNewTodo = (todo) => {
        const action = {
            type: addTodoAction,
            payload: todo
        }
        dispatch(action);
    }

    const handleDeleteTodo = (todo) => {
        const action = {
            type: deleteTodoAction,
            payload: todo
        };
        dispatch(action);
    }

    const handleToggleTodo = (todo) => {
        const action = {
            type: toggleTodoAction,
            payload: todo
        };
        dispatch(action);
    }

    return {
        todos,
        todosCount,
        pendingTodosCount,
        handleNewTodo,
        handleDeleteTodo,
        handleToggleTodo
    }
};

