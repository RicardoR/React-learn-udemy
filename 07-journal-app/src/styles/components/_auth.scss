.auth__main {
    display: flex;
    justify-content: center;
    align-items: center;
    background-color: $primary;
    margin: 0px;
    height: 100vh;
    width: 100vw;
}

.auth__box-container {
    background-color: white;
    box-shadow: 0px 3px $dark-grey;
    border-radius: 2px;
    padding: 20px;
    width: 250px;
}

.auth__title {
    color: darken($primary, 20%);
}

.auth__input {
    color: $dark-grey;
    border: 0px;
    border-bottom: 1px solid $light-grey;
    font-size: 16px;
    margin-bottom: 10px;
    height: 20px;
    width: 100%;
    transition: border-bottom .3s ease;

    &:focus {
        outline: none;
        border-bottom: 1px solid $primary;
    }
}

