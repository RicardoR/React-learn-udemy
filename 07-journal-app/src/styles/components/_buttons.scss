.btn {
    background-color: transparent;
    border: none;
    color: white;
    cursor: pointer;
    font-size: 12px;
    padding: 7px 10px 7px 10px;
    transition: color 0.3s ease-in-out;
    &:focus {
        outline: none;
    }

    &:hover {
        color: darken($color: white, $amount: 20);
    }
}

.btn-primary {
    background-color: $primary;
    border-radius: 2px;
    box-shadow: 0 3px 4px 0 rgba(0, 0, 0, 0.25);
    transition: box-shadow .3s ease;
    &:hover {
        box-shadow: 0 0 6px #4285f4;
        background-color: darken($color: $primary, $amount: 15);
    }
   
    &:active {
        background: #1669f2;
    }
    &:disabled {
        background-color: lighten($color: $primary, $amount: 15);
    }
}

.btn-block {
    display: block;
    width: 100%;
}