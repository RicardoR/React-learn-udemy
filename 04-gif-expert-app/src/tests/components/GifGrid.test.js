import React from 'react';
import { shallow } from 'enzyme';
import '@testing-library/jest-dom';
import '@testing-library/react';
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';

// Attention!! Mocking a service return data
jest.mock('../../hooks/useFetchGifs');

describe('<GifGrid />', () => {
    const category = 'Avengers'
    const gifs = [
      {
        id: 'wertret56456456',
        url: 'https://localhost/avengers.jpg',
        title: 'Avengers',
      },
    ];

    useFetchGifs.mockReturnValue({
      loading: true,
      data: gifs,
    });

    const wrapper = shallow(<GifGrid category={category} />);

    test('should create', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should have a title with the category', () => {
        expect(wrapper.find('h3').text().trim()).toBe(category);
    });

    test('should display items when images are loaded', () => {
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length);
    });
});
