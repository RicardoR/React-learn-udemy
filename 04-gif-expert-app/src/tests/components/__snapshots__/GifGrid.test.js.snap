// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GifGrid /> should create 1`] = `
<Fragment>
  <h3>
    Avengers
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="wertret56456456"
      key="wertret56456456"
      title="Avengers"
      url="https://localhost/avengers.jpg"
    />
  </div>
</Fragment>
`;
